### PR TITLE
ci: Deploy pages

### PR DESCRIPTION
**Pull request by ggera**
## fixes KILTProtocol/ticket#3583
Deploys master as default docs branch then each tag and develop are put inside the root directory
Succeful deployment can be seen here https://github.com/KILTprotocol/kilt-node/actions/runs/10595010314

## Metadata Diff to Develop Branch

<details>
<summary>Peregrine Diff</summary>

```
```

</details>

<details>
<summary>Spiritnet Diff</summary>

```
```

</details>

## Checklist:

- [ ] I have verified that the code works
  - [ ] No panics! (checked arithmetic ops, no indexing `array[3]` use `get(3)`, ...)
- [ ] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [ ] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [ ] I have documented the changes (where applicable)
    * Either PR or Ticket to update [the Docs](https://github.com/KILTprotocol/docs)
    * Link the PR/Ticket here

### CODE REVIEW

**Comment by ChrisChinchilla** [comment]
@ggera Will this replace the existing job?

**Comment by ggera** [comment]
> @ggera Will this replace the existing job?

Thanks for catching that, meant to update exsiting action

**Review by ChrisChinchilla — DISMISSED** [review]

**Comment by ntn-x2** [comment]
@ggera I rebased the CI branch on top of latest `develop`. Can you do the same for this branch on top of the CI branch? The diff set seems weird.

File: .github/workflows/publish-rustdocs.yml [hunk 1]
```diff
@@ -43,12 +49,51 @@ jobs:
             -v "${GITHUB_WORKSPACE}:/workspace" \
             -v "${HOME}/.cargo:/root/.cargo" \
             -w /workspace \
-            -e RUSTDOCFLAGS='-D warnings' \
```
Comment (ntn-x2) [inline .github/workflows/publish-rustdocs.yml:46]: Isn't this env propagated inside the Docker container and inside the `bash` program?

File: .github/workflows/publish-rustdocs.yml [hunk 2]
```diff
@@ -43,12 +49,51 @@ jobs:
...
             paritytech/ci-unified:bullseye-1.74.0 \
-            bash -c "cargo doc --all-features --no-deps --locked && chown -R $(id -u):$(id -g) ${{ env.DOCS_TARGET }}"
+            bash -c "
+              RUSTDOCFLAGS='-D warnings' cargo doc --all-features --no-deps --locked && \
+              mv $DOCS_TARGET/help.html $DOCS_TARGET/index.html && chown -R $(id -u):$(id -g) $DOCS_TARGET"
 
-      - name: Deploy to GitHub Pages
+      - name: Push to gh-pages for master branch
```
[the remark this answers is not in the document]
Comment (ChrisChinchilla) [reply .github/workflows/publish-rustdocs.yml:57]: It was a single job before I think, what was the reason for splitting it?

[the remark this answers is not in the document]
Comment (ggera) [reply .github/workflows/publish-rustdocs.yml]: following the best practice: `We recommend this action to be used in a dedicated job:`
https://github.com/actions/deploy-pages. Moreover the `build` job is done the default src code branch where as the `deployment` job checks out generated docs branch which is gh-pages

File: .github/workflows/publish-rustdocs.yml [hunk 3]
```diff
@@ -43,12 +49,51 @@ jobs:
             -v "${GITHUB_WORKSPACE}:/workspace" \
             -v "${HOME}/.cargo:/root/.cargo" \
             -w /workspace \
-            -e RUSTDOCFLAGS='-D warnings' \
             paritytech/ci-unified:bullseye-1.74.0 \
-            bash -c "cargo doc --all-features --no-deps --locked && chown -R $(id -u):$(id -g) ${{ env.DOCS_TARGET }}"
+            bash -c "
+              RUSTDOCFLAGS='-D warnings' cargo doc --all-features --no-deps --locked && \
+              mv $DOCS_TARGET/help.html $DOCS_TARGET/index.html && chown -R $(id -u):$(id -g) $DOCS_TARGET"
```
[the remark this answers is not in the document]
Comment (ggera) [reply .github/workflows/publish-rustdocs.yml:61]: `${{ }} `for `yml vars`  whereas env variables inside a `bash code block` could accessed with just` $ENV`

File: .github/workflows/publish-rustdocs.yml (hunk 1 above)
Comment (ggera) [reply .github/workflows/publish-rustdocs.yml:46]: moved it up

File: .github/workflows/publish-rustdocs.yml [hunk 4]
```diff
@@ -42,13 +53,50 @@ jobs:
...
             -v "${GITHUB_WORKSPACE}:/workspace" \
             -v "${HOME}/.cargo:/root/.cargo" \
-            -w /workspace \
             -e RUSTDOCFLAGS='-D warnings' \
+            -w /workspace \
             paritytech/ci-unified:bullseye-1.74.0 \
-            bash -c "cargo doc --all-features --no-deps --locked && chown -R $(id -u):$(id -g) ${{ env.DOCS_TARGET }}"
+            bash -c "
+              cargo doc --all-features --no-deps --locked && \
+              mv $DOCS_TARGET/help.html $DOCS_TARGET/index.html && chown -R $(id -u):$(id -g) $DOCS_TARGET"
```
Comment (ntn-x2) [inline .github/workflows/publish-rustdocs.yml:61]: Just move the third command on its own line as well.

File: .github/workflows/publish-rustdocs.yml [hunk 5]
```diff
@@ -42,13 +53,50 @@ jobs:
...
+        if: github.ref_name != 'master'
+        run: echo "TARGET_FOLDER=${{ github.ref_name }}" >> $GITHUB_ENV
+
+      - name: Push to gh-pages branch
         uses: JamesIves/github-pages-deploy-action@v4
         with:
-          branch: gh-pages-${{ github.ref_name }}
+          branch: ${{ env.BRANCH }}
           folder: ${{ env.DOCS_TARGET }}
+          target-folder: ${{ env.TARGET_FOLDER }}
```
Comment (ntn-x2) [inline .github/workflows/publish-rustdocs.yml:72]: What happens if this is then not specified, i.e., we are on `master`?

File: .github/workflows/publish-rustdocs.yml (hunk 4 above)
Comment (ggera) [reply .github/workflows/publish-rustdocs.yml]: Done

File: .github/workflows/publish-rustdocs.yml (hunk 5 above)
Comment (ggera) [reply .github/workflows/publish-rustdocs.yml:73]: `target-folder=''` hence will be deployed to default root direcotry

**Review by ntn-x2 — APPROVED** [review]

**Pull request merged**